pkg/istio-agent/health/health_check.go

```diff
@@ -0,0 +1,141 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package health
+
+import (
+	"time"
+
+	"istio.io/api/networking/v1alpha3"
+)
+
+const (
+	HealthInfoTypeURL string = "type.googleapis.com/istio.v1.HealthInformation"
+)
+
+type WorkloadHealthChecker struct {
+	config applicationHealthCheckConfig
+	prober Prober
+}
+
+// internal field purely for convenience
+type applicationHealthCheckConfig struct {
+	InitialDelay   time.Duration
+	ProbeTimeout   time.Duration
+	CheckFrequency time.Duration
+	SuccessThresh  int
+	FailThresh     int
+}
+
+type ProbeEvent struct {
+	Healthy          bool
+	UnhealthyStatus  int32
+	UnhealthyMessage string
+}
+
+func NewWorkloadHealthChecker(cfg *v1alpha3.ReadinessProbe) *WorkloadHealthChecker {
+	if cfg == nil {
+		return &WorkloadHealthChecker{
+			config: applicationHealthCheckConfig{},
+			prober: &NoOpProber{},
+		}
+	}
+	var prober Prober
+	switch healthCheckMethod := cfg.HealthCheckMethod.(type) {
+	case *v1alpha3.ReadinessProbe_HttpGet:
+		prober = &HTTPProber{Config: healthCheckMethod.HttpGet}
+	case *v1alpha3.ReadinessProbe_TcpSocket:
+		prober = &TCPProber{Config: healthCheckMethod.TcpSocket}
+	case *v1alpha3.ReadinessProbe_Exec:
+		prober = &ExecProber{Config: healthCheckMethod.Exec}
+	default:
+		prober = &NoOpProber{}
+	}
+
+	return &WorkloadHealthChecker{
+		config: applicationHealthCheckConfig{
+			InitialDelay:   time.Duration(cfg.InitialDelaySeconds) * time.Second,
+			ProbeTimeout:   time.Duration(cfg.TimeoutSeconds) * time.Second,
+			CheckFrequency: time.Duration(cfg.PeriodSeconds) * time.Second,
+			SuccessThresh:  int(cfg.SuccessThreshold),
+			FailThresh:     int(cfg.FailureThreshold),
+		},
+		prober: prober,
+	}
+}
+
+// PerformApplicationHealthCheck Performs the application-provided configuration health check.
+// Instead of a heartbeat-based health checks, we only send on a health state change, and this is
+// determined by the success & failure threshold provided by the user.
+func (w *WorkloadHealthChecker) PerformApplicationHealthCheck(notifyHealthChange chan *ProbeEvent, quit chan struct{}) {
```

Comment:
if there is no health check config set (and remember, this will be the case for 100% of users at first, so we need to optimize for this), can we just exit early? Or don't even call this function to begin with